test/integration/weighted_cluster_integration_test.cc

```diff
@@ -0,0 +1,190 @@
+            cluster_with_header->set_cluster_header(
+                absl::StrFormat(Envoy::RepickClusterFilter::ClusterHeaderName, i));
+            cluster_with_header->mutable_weight()->set_value(weights[i]);
+          }
+
+          hcm.mutable_route_config()
+              ->mutable_virtual_hosts(0)
+              ->mutable_routes(0)
+              ->mutable_route()
+              ->mutable_weighted_clusters()
+              ->mutable_total_weight()
+              ->set_value(std::accumulate(weights.begin(), weights.end(), 0UL));
+        });
+
+    HttpIntegrationTest::initialize();
+  }
+
+  const std::vector<uint64_t>& getDefaultWeights() { return default_weights_; }
+
+  void sendRequestAndValidateResponse(const std::vector<uint64_t>& upstream_indices) {
+    // Create a client aimed at Envoy’s default HTTP port.
+    codec_client_ = makeHttpConnection(makeClientConnection((lookupPort("http"))));
+
+    // Create some request headers.
+    Http::TestRequestHeaderMapImpl request_headers{{":method", "GET"},
+                                                   {":path", "/test/long/url"},
+                                                   {":scheme", "http"},
+                                                   {":authority", "host"}};
+
+    // Send the request headers from the client, wait until they are received
+    // upstream. When they are received, send the default response headers from
+    // upstream and wait until they are received at by client.
+    IntegrationStreamDecoderPtr response = sendRequestAndWaitForResponse(
+        request_headers, 0, default_response_headers_, 0, upstream_indices);
+
+    // Verify the proxied request was received upstream, as expected.
+    EXPECT_TRUE(upstream_request_->complete());
+    EXPECT_EQ(0U, upstream_request_->bodyLength());
+    // Verify the proxied response was received downstream, as expected.
+    EXPECT_TRUE(response->complete());
+    EXPECT_EQ("200", response->headers().getStatusValue());
+    EXPECT_EQ(0U, response->body().size());
+
+    // Perform the clean-up.
+    cleanupUpstreamAndDownstream();
+  }
+
+private:
+  std::vector<uint64_t> default_weights_;
+};
+
+TEST_F(WeightedClusterIntegrationTest, SteerTrafficToOneClusterWithName) {
+  setDeterministicValue(0);
+  initializeConfig(getDefaultWeights());
+
+  // The expected destination cluster upstream is index 0 since the selected
+  // value is set to 0 indirectly via `setDeterministicValue(0)`.
+  sendRequestAndValidateResponse({0});
+}
+
+TEST_F(WeightedClusterIntegrationTest, SteerTrafficToOneClusterWithHeader) {
+  const std::vector<uint64_t>& default_weights = getDefaultWeights();
+  uint64_t deterministric_value = std::accumulate(
+      default_weights.begin(), default_weights.begin() + TotalUpstreamClusterWithNameCount, 0UL);
+  setDeterministicValue(deterministric_value);
+  initializeConfig(default_weights);
+
+  // The expected destination cluster upstream is index
+  // `TotalUpstreamClusterWithNameCount`.
```

Comment:
I would probably have written this test has having `WeightedClusterIndex = 0` and `WeightedClusterHeaderIndex = 1`, with one cluster designated as the one that we hit via the original cluster weighting mechanism, the other one being targeted via a cluster header. I think that would make these tests a bit easier to follow instead of having to just work with generic cluster indices.

Reply:
While generic code can be great, I'm hesitant to merge in generic code that is only used with one configuration. While generic code is great, it does bring with it complexity that makes this code harder to understand for everyone reading it going forward. I would suggest simplifying the test filter to only doing a single, simple thing and then have the integration test itself use well named clusters. 

It's great that this pattern let you verify things locally, but I don't think we should be checking in code that is more complex than necessary for what is actually run in CI

Reply:
Thanks for laying out the thoughts!

I took a closer look at code.  After changing the `random number of cluster` to `fixed number (i.e. 2) cluster`, I feel most code will be same except few things: 1) we can have fixed cluster string name e.g. `cluster_header_0`  rather than `absl::StrFormat(ClusterHeaderName, i)` to construct to name 2) We have simpler pattern of clusters (just 2 clusters) and avoid some math calculation for cluster array.

I think we will still have some global variable values rather than using magic number `1` or `2` multiple places as that might not be good for readability.  

So just my 2 cents, I don't feel there are lots of code/things would be simplified by changing to fixed number cluster . Also, the flexibility and configurability outweigh the things could be simplified here. More comments may helps here

I added some comments, tweak some variable names and add  local variables in the hope of clarifying the code. Please let me know your thoughts. Thanks

Reply:
Per offline discussion, i tweaked the variable names and code comments.

I also tweaked the cluster array config a bit: change from 2 (1 `name`+ 1 `cluster_header`) to (1 `name` + 2 `cluster_header`) .

PTAL! Thanks.